File: Python/Stack/ValidParenthesis.py
# define a function to check for valid parenthesis

def is_valid_parenthesis(s):
    stack = []

    for char in s:

        # opening the bracket
        if char in '({[':
            stack.append(char)
        
        # closing the bracket

        if char in ')}]':
            
            # closing bracket without opening
            if not stack:
                return False
            
            # else pop an item check for matching
            top = stack.pop()

            if (top == '(' and char != ')') or \
                (top == '{' and char != '}') or \
                (top == '[' and char != ']'):
                return False
            
    # if an opening bracket without closing
    return len(stack) == 0


# example
s = '{()}[]'

if is_valid_parenthesis(s):
    print("Valid Parenthesis")
else:
    print("Invalid Parenthesis")


            

File: Javascript/Stack/ValidParenthesis.js
const isValidParenthesis = (s) => {
    let stack = []

    for (let i=0; i<stack.length; i++) {
        
        if (s[i] === '(' || s[i] === '{' || s[i] === '[') {
            stack.push(stack[i])
        } else {
            // if its's a closing bracket, check if the stack is non-empty
            // and if the top of the stack is a matching opening bracket
            if (stack.length > 0 && 
                ((stack[stack.length -1] === '(' && s[i] === ')')) || 
                ((stack[stack.length -1] === '{' && s[i] === '}')) || 
                ((stack[stack.length -1] === '[' && s[i] === ']'))
            ) {
                stack.pop() // Pop the matching opening bracket
            } else {
                return false  // Unmatched closing bracket
            }
        }
    }

    // If stack is empty, return true (balanced), otherwise false
    return stack.length === 0;
}

let s = "{()}[]";

console.log(isValidParenthesis(s) ? "Valid Parenthesis" : "Invalid Parenthesis")

